src/io/parser/types.rs

```diff
@@ -100,6 +100,9 @@ pub enum ParseError {
         r#"SPARQL data source for predicate "{0}" has arity {1}, but {2} variables are given"#
     )]
     SparqlSourceInvalidArity(String, usize, usize),
+    /// A data source declarations types do not match its arity.
+    #[error(r#"Data source for predicate "{0}" has arity "{1}" but {2} types are given."#)]
+    SourceInvalidTypeArity(String, usize, usize),
```

Comment:
`declaration's`, but we should rather make this something that cannot occur, by only allowing either of arity and types to be given, not both.